go/engine/selfprovision.go

```diff
@@ -0,0 +1,295 @@
+package engine
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/keybase/client/go/libkb"
+)
+
+type SelfProvisionEngine struct {
+	libkb.Contextified
+	DeviceName     string
+	result         error
+	lks            *libkb.LKSec
+	User           *libkb.User
+	perUserKeyring *libkb.PerUserKeyring
+}
+
+// If a device is cloned, we can provision a new device from the current device
+// to get out of the cloned state.
+func NewSelfProvisionEngine(g *libkb.GlobalContext, deviceName string) *SelfProvisionEngine {
+	return &SelfProvisionEngine{
+		Contextified: libkb.NewContextified(g),
+		DeviceName:   deviceName,
+	}
+}
+
+func (e *SelfProvisionEngine) Name() string {
+	return "SelfProvision"
+}
+
+func (e *SelfProvisionEngine) Prereqs() Prereqs {
+	return Prereqs{}
+}
+
+func (e *SelfProvisionEngine) RequiredUIs() []libkb.UIKind {
+	return []libkb.UIKind{
+		libkb.ProvisionUIKind,
+		libkb.LogUIKind,
+		libkb.SecretUIKind,
+		libkb.LoginUIKind,
+	}
+}
+
+func (e *SelfProvisionEngine) Run(m libkb.MetaContext) (err error) {
+	m.G().LocalSigchainGuard().Set(m.Ctx(), "selfProvision")
+	defer m.G().LocalSigchainGuard().Clear(m.Ctx(), "selfProvision")
+	defer m.CTrace("SelfProvisionEngine#Run", func() error { return err })()
+
+	if d, err := libkb.GetDeviceCloneState(m); err != nil {
+		return err
+	} else if !d.IsClone() {
+		return fmt.Errorf("to self provision, you must be a cloned device")
+	}
+
+	// transaction around config file
+	tx, err := e.G().Env.GetConfigWriter().BeginTransaction()
+	if err != nil {
+		return err
+	}
+
+	// If we abort, we need to revert to the old active device and user config
+	// state
+	uvOld, deviceIDOld, deviceNameOld, sigKeyOld, encKeyOld := e.G().ActiveDevice.AllFields()
+	// Pass the UV here so the passphrase stream is cached on the provisional
+	// login context
+	m = m.WithNewProvisionalLoginContextForUserVersionAndUsername(uvOld, e.G().Env.GetUsername())
+
+	// From this point on, if there's an error, we abort
+	// the transaction.
+	defer func() {
+		if tx != nil {
+			tx.Abort()
+		}
+		if err == nil {
+			m = m.CommitProvisionalLogin()
+		} else if e.User != nil {
+			m.CDebugf("Error in self provision, reverting to original active device: %v", err)
+			m = m.WithGlobalActiveDevice()
+			salt, err := e.User.GetSalt()
+			if err != nil {
+				m.CDebugf("unable to GetSalt: %v", err)
+				return
+			}
+			// Atomically swap to the new config and active device
+			if err := m.SwitchUserNewConfigActiveDevice(uvOld, e.User.GetNormalizedName(), salt,
```

Comment:
It's a slight bummer we have to explicitly rollback.  We've avoided rollbacks in other flows by just writing the configuration file as the last thing. Is this not possible for this flow?

Reply:
i was having some trouble with this so added it explicitly, let me do some digging as to why the transaction on the config isn't working

Reply:
what if we don't write the config file in DeviceWrap and just do it at the very end, if all worked?  Does that screw anything else up?